### ISSUE TITLE
空洞卷积比较耗时

### ISSUE THREAD

**Issue by bobzhang123** [alibaba/MNN#172]
我这边进行的一个任务是mobilenetv2+aspp架构，目前在测试的过程中，发现空洞卷积（aspp模块）比较耗时，请问你们那边有对这块儿进行优化嘛？
  tf的代码段如下：
    # aspp
    with tf.variable_scope(None, 'ASPP'):
        aspp_0 = conv2d(net, 96, 1, 1, 1, 1, name='aspp_0')
        filter = tf.constant(value=1, shape=[3, 3, 96, 96], dtype=tf.float32)
        aspp_1 = tf.nn.atrous_conv2d(net, filter, 2, padding='SAME', name='aspp_1')
        aspp_2 = tf.nn.atrous_conv2d(net, filter, 4, padding='SAME', name='aspp_2')
        aspp_3 = tf.nn.atrous_conv2d(net, filter, 6, padding='SAME', name='aspp_3')
        avg_pooling = tf.nn.avg_pool(net, [1, 3, 3, 1], [1, 1, 1, 1], padding='SAME', name='avg_pooling')
        aspp_concat = tf.concat([aspp_0, aspp_1, aspp_2, aspp_3, avg_pooling], 3, name='aspp_concat')
        aspp_all = separable_conv(aspp_concat, 3, 96, 1, name='aspp_all')

转化为MNN后，利用MNNV2Basic.out在PC上测试的时间消耗如下：
              MV2_ASPP2/res5_4/res5_4/pw/pw/Conv2D run 10 average cost 0.929900 ms, 0.792 %, FlopsRate: 3.130 %
              MV2_ASPP2/res6_1/res6_1/pw/pw/Conv2D run 10 average cost 0.943300 ms, 0.804 %, FlopsRate: 3.130 %
              MV2_ASPP2/res3_1/res3_1/pw/pw/Conv2D run 10 average cost 1.102500 ms, 0.940 %, FlopsRate: 3.130 %
           MV2_ASPP2/conv1_1/conv1_1/conv2d/Conv2D run 10 average cost 1.222500 ms, 1.042 %, FlopsRate: 1.761 %
MV2_ASPP2/res6_1/res6_1/pw_linear/pw_linear/Conv2D run 10 average cost 1.294300 ms, 1.103 %, FlopsRate: 4.696 %
         MV2_ASPP2/ASPP/aspp_all/aspp_all/aspp_all run 10 average cost 1.791100 ms, 1.526 %, FlopsRate: 5.869 %
                             MV2_ASPP2/ASPP/aspp_1 run 10 average cost 15.434401 ms, 13.153 %, FlopsRate: 10.565 %
                             MV2_ASPP2/ASPP/aspp_2 run 10 average cost 19.099400 ms, 16.277 %, FlopsRate: 11.269 %
                             MV2_ASPP2/ASPP/aspp_3 run 10 average cost 53.419098 ms, 45.524 %, FlopsRate: 11.093 %
Avg= 117.342606 ms, min= 115.272003 ms, max= 121.570000 ms
aspp模块的总耗时为13.15%+16.3%+45.5%=75%，时间占比约为75%。

同样的模型，转化为coreML后，在iphoneXR上运行时间达到了33ms左右，而在一加（晓龙845）上的速度仅仅为200ms左右。

目前定位到的问题是空洞卷积的问题，是否为MNN上空洞卷积实现的效率过低？如有能否对这部分进行优化？

**Comment by jxt1234** [comment]
1､先确认一下是不是空洞卷积（即 mnn convolution 的 dilate 值不为 1）
2、目前MNN CPU 上空洞卷积是走的滑窗算法（也是汇编实现），相对 MatMul 和 Winograd 方式效率低一些，如果特征尺寸较大，会有较大的性能劣势，正在优化中
3、可以先用 MNN 的 GPU ，opencl 和 vulkan backend 对 dilate 的实现是 im2col + MatMul ，相对高效一些

**Comment by bobzhang123** [comment]
@jxt1234 谢谢你的回复！
我这边aspp架构空洞卷积的dilate分别为2、4、6，我们已经把算法部署在了移动端，实测发现不管是CPU还是GPU（主要测试opencl），都比较慢，速度差别不算大。

**Comment by YIFanH** [comment]
@bobzhang123 这块移植到端上 耗时如何？
